src/auth/integration-tests/src/lib.rs

```diff
@@ -78,6 +80,57 @@ pub async fn service_account() -> anyhow::Result<()> {
     Ok(())
 }
 
+pub async fn impersonated() -> anyhow::Result<()> {
+    let project = std::env::var("GOOGLE_CLOUD_PROJECT").expect("GOOGLE_CLOUD_PROJECT not set");
+
+    // Create a SecretManager client. When running on GCB, this loads MDS
+    // credentials for our `integration-test-runner` service account.
+    let client = SecretManagerService::builder().build().await?;
+
+    // Load the service account json that will be the source credential
+    let response = client
+        .access_secret_version()
+        .set_name(format!(
+            "projects/{project}/secrets/test-sa-creds-json/versions/latest"
+        ))
+        .send()
+        .await?;
+    let source_sa_json = response
+        .payload
+        .expect("missing payload in test-sa-creds-json response")
+        .data;
+
+    let source_sa_json: serde_json::Value = serde_json::from_slice(&source_sa_json)?;
+
+    let source_sa_creds = ServiceAccountCredentialsBuilder::new(source_sa_json).build()?;
+
+    let impersonated_creds =
+        ImpersonatedCredentialsBuilder::from_source_credentials(source_sa_creds)
+            .with_target_principal("impersonation-target@rust-auth-testing.iam.gserviceaccount.com")
+            .build()?;
+
+    let client = SecretManagerService::builder()
+        .with_credentials(impersonated_creds)
+        .build()
+        .await?;
+
+    // Access a secret, which only this principal has permissions to do.
+    let response = client
+        .access_secret_version()
+        .set_name(format!(
+            "projects/{project}/secrets/impersonation-target-secret/versions/latest"
+        ))
+        .send()
+        .await?;
+    let secret = response
+        .payload
+        .expect("missing payload in impersonation-target-secret response")
+        .data;
+    assert_eq!(secret, "impersonated_secret_value");
+
```

Comment:
optional: consider also trying to use the impersonating principal directly to access the secret and verify that that fails.

On my last team, we claimed we had implemented impersonation, when actually we were using the impersonating SA principal directly.

The impersonating SA used in our tests had the same permissions as the impersonated SA. 🤦

Reply:
This is less important if the test SA configuration is checked into the repo